File: algorithms/php/1608.SpecialArrayWithXElementsGreaterThanorEqualX.php
<?php

class Solution
{

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function specialArray($nums)
    {
        sort($nums);

        if (count($nums) < $nums[0]) {
            return count($nums);
        }

        $l = 0;
        $r = count($nums) - 1;

        while ($l <= $r) {
            $m = $l + (($r - $l) >> 1);
            $x = count($nums) - $m;

            if ($nums[$m] >= $x) {
                if ($nums[$m - 1] < $x) {
                    return $x;
                }
                $r = $m - 1;
            } else {
                $l = $m + 1;
            }
        }

        return -1;
    }
}

File: algorithms/cpp/1608.SpecialArrayWithXElementsGreaterThanorEqualX.cpp
class Solution {
public:
    int count(vector<int>& nums, int m) {
        int cnt = 0;
        for (int i : nums) {
            if (i >= m) {
                cnt++;
            }
        }

        return cnt;
    }


    int specialArray(vector<int>& nums) {
        int l = 0;
        int r = nums.size();

        while (l <= r) {
            int m = l + (r - l) / 2;
            int cnt = count(nums, m);

            if (cnt == m) {
                return m;
            }

            if (m < cnt) {
                l = m + 1;
            }

            if (m > cnt) {
                r = m - 1;
            }
        }

        return - 1;
    }

};
